### ISSUE TITLE
Issue Tracker: Refactor highlight provider

### ISSUE THREAD

**Issue by theol0403** [vscode-neovim/vscode-neovim#1264]
Currently, highlight provider is very bloated and buggy, especially to do with tabs in the file. Should be refactored.

- [x] https://github.com/vscode-neovim/vscode-neovim/issues/519
- [ ] https://github.com/vscode-neovim/vscode-neovim/issues/329
- [ ] https://github.com/vscode-neovim/vscode-neovim/issues/695
- [x] https://github.com/vscode-neovim/vscode-neovim/issues/1004
- [x] https://github.com/vscode-neovim/vscode-neovim/issues/1144
- [ ] https://github.com/vscode-neovim/vscode-neovim/issues/1205
- [ ] https://github.com/vscode-neovim/vscode-neovim/issues/1191

**Comment by theol0403** [comment]
@justinmk would love ideas here.

## Problem 1: syntax off
Many neovim syntax-aware features (see linked issues) are broken, because `syntax off` is set. Ideally, we have `syntax on`.

Why is `syntax off``? Because we don't want to render neovim syntax highlights, which would look bad and spam the highlight provider. 

Why can't we ignore the syntax groups? Because many extensions will `hi Link` to syntax groups, so ignoring will break them. Also, even if they are ignored, they will clog up rpc because they are ignored on vscode side. 

### Potential Solutions

Ideally, we want a way to have `syntax on`, *without actually doing syntax highlighting*. This will support plugins linking to syntax groups, but avoid syntax highlighting and clogging rpc. 

However, I'm not sure what options are available to do this.

Another solution would be to somehow break all the `:hi link` to render as their own group. 

## Problem 1: syntax group ignoring
The way syntax group ignoring is currently done is messy and buggy. 

Lets follow an example. Say we want to ignore `MatchParen`. Neovim will create a highlight id containing `MatchParen`, and report it with `hl_attr_define` along with all decoration attributes. **Currently, vscode-neovim indexes highlights by group name, not id**. Thus, it will store `MatchParen` in a map, and also add the highlight id to a blacklist. 

The reason it still bothers storing the decorator is in the event that external buffers want to be used, with full syntax highlighting. 

Now, whenever `grid_line` reports a highlight id matching the blacklist, it will ignore it.

But, what happens when there are multiple syntax groups on a cell? For example, `MatchParen` + `Visual`. Neovim will calculate the combination and produce a new highlight id with the calculated values. But how do we store this, and how do we ignore `MatchParen` when it is combined like this? 

As of https://github.com/vscode-neovim/vscode-neovim/pull/1258, what it does is merge the names together and store a fake syntax group called `MatchParenVisual`. Then, when a cell is highlighted, it will check if the highlight group name associated with the highlight id contains a blacklisted group. It will then strip it out. Thus, the flow is `grid_line` -> `id` -> `MatchParenVisual` -> `Visual` -> decoration. 

TLDR; These are hacks, ignoring from vscode side is messy and bloated, we should be indexing by id, and cleaning up rpc communication.

### Potential solutions
Instead of ignoring groups from vscode side, we ignore them on the neovim side, by setting `hi MatchParen NONE`. This cleans up rpc and solves the merged highlights problem. I can't think of many problems with this atm.
